crates/sui-indexer-restorer/src/snapshot.rs

```diff
@@ -0,0 +1,214 @@
+use sui_indexer_alt::schema::obj_info;
+use sui_snapshot::{
+    reader::{download_bytes, LiveObjectIter, StateSnapshotReaderV1},
+    FileMetadata,
+};
+use sui_storage::object_store::ObjectStoreGetExt;
+
+use crate::Args;
+
+pub type DigestByBucketAndPartition = BTreeMap<u32, BTreeMap<u32, [u8; 32]>>;
+
+pub struct SnapshotRestorer {
+    pub restore_args: Args,
+    pub next_checkpoint_after_epoch: u64,
+    pub snapshot_reader: StateSnapshotReaderV1,
+    pub db: Db,
+}
+
+impl SnapshotRestorer {
+    pub async fn new(args: &Args, next_checkpoint_after_epoch: u64) -> Result<Self, Error> {
+        let remote_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::S3),
+            aws_endpoint: Some(args.endpoint.clone()),
+            aws_virtual_hosted_style_request: true,
+            object_store_connection_limit: args.concurrency,
+            no_sign_request: true,
+            ..Default::default()
+        };
+
+        let local_path = PathBuf::from(&args.snapshot_local_dir);
+        let snapshot_dir = local_path.join("snapshot");
+        let local_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::File),
+            directory: Some(snapshot_dir.clone().to_path_buf()),
+            ..Default::default()
+        };
+
+        let m = MultiProgress::new();
+        let snapshot_reader = StateSnapshotReaderV1::new(
+            args.start_epoch,
+            &remote_store_config,
+            &local_store_config,
+            usize::MAX,
+            NonZeroUsize::new(args.concurrency).unwrap(),
+            m,
+            true, // skip_reset_local_store
+        )
+        .await?;
+
+        let database_url = Url::parse(&args.database_url)?;
+        let db_args = DbArgs::with_url(database_url);
+        let db = Db::new(db_args).await?;
+
+        Ok(Self {
+            restore_args: args.clone(),
+            snapshot_reader,
+            db,
+            next_checkpoint_after_epoch,
+        })
+    }
+
+    pub async fn restore(&mut self) -> Result<(), Error> {
+        info!(
+            "Starting snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        let (sha3_digests, num_part_files) = self.snapshot_reader.compute_checksum().await?;
+        let (_abort_handle, abort_registration) = AbortHandle::new_pair();
+        let (input_files, epoch_dir, remote_object_store, _concurrency) =
+            self.snapshot_reader.export_metadata().await?;
+        let owned_input_files: Vec<(u32, (u32, FileMetadata))> = input_files
+            .into_iter()
+            .map(|(bucket, (part_num, metadata))| (*bucket, (part_num, metadata.clone())))
+            .collect();
+
+        self.restore_object_infos(
+            abort_registration,
+            owned_input_files,
+            epoch_dir,
+            remote_object_store,
+            sha3_digests,
+            num_part_files,
+            self.restore_args.concurrency,
```

Comment:
Is this ever not `self.restore_args.concurrency`? This function has access to `self` so presumably it can fetch this parameter itself as well.